src/scripting/lua_widget_attributes.cpp

```diff
@@ -0,0 +1,527 @@
+	w.set_value(value);
+}
+
+WIDGET_GETTER("value_compat,percentage", int, gui2::progress_bar)
+{
+	return w.get_percentage();
+}
+
+WIDGET_SETTER("value_compat,percentage", int, gui2::progress_bar)
+{
+	w.set_percentage(value);
+}
+
+WIDGET_GETTER("value_compat,selected_item_path", std::vector<int>, gui2::tree_view)
+{
+	auto res = w.selected_item()->describe_path();
+	for(int& a : res) { ++a;}
+	return res;
+}
+
+WIDGET_GETTER("path", std::vector<int>, gui2::tree_view_node)
+{
+	auto res = w.describe_path();
+	for(int& a : res) { ++a;}
+	return res;
+}
+
+WIDGET_SETTER("value_compat,unfolded", bool, gui2::tree_view_node)
+{
+	if(value) {
+		w.unfold();
+	} else {
+		w.fold();
+	}
+}
+
+WIDGET_SETTER("value_compat,unit", lua_index_raw, gui2::unit_preview_pane)
+{
+	if(const unit_type* ut = luaW_tounittype(L, value.index)) {
+		w.set_displayed_type(*ut);
+	} else if(unit* u = luaW_tounit(L, value.index)) {
+		w.set_displayed_unit(*u);
+	} else {
+		luaW_type_error(L, value.index, "unit or unit type");
+	}
+}
+
+WIDGET_GETTER("item_count", int, gui2::multi_page)
+{
+	return w.get_page_count();
+}
+
+WIDGET_SETTER("use_markup", bool, gui2::styled_widget)
+{
+	w.set_use_markup(value);
+}
+
+//TODO: while i think this shortcut is useful, i'm not that happy about
+//      the name since  it changes 'label' and not 'text', the first one
+//      is the label that is part of most widgets (like checkboxes), the
+//      later is sepcific to input textboxes.
+WIDGET_SETTER("marked_up_text", t_string, gui2::styled_widget)
+{
+	w.set_use_markup(true);
+	w.set_label(value);
+}
+
+WIDGET_SETTER("enabled", bool, gui2::styled_widget)
+{
+	w.set_active(value);
+}
+
+WIDGET_SETTER("tooltip", t_string, gui2::styled_widget)
+{
+	w.set_tooltip(value);
+}
+
+
+WIDGET_SETTER("callback", lua_index_raw, gui2::widget)
+{
+	if(!luaW_getglobal(L, "gui", "widget", "set_callback")) {
+		ERR_LUA << "gui.widget.set_callback didnt exist\n";
+	}
+	luaW_pushwidget(L, w);
+	lua_pushvalue(L, value.index);
+	lua_call(L, 2, 0);
+}
+
+WIDGET_SETTER("visible", lua_index_raw, gui2::styled_widget)
+{
+
+	typedef gui2::styled_widget::visibility visibility;
+
+	visibility flag = visibility::visible;
+
+	switch(lua_type(L, value.index)) {
+		case LUA_TBOOLEAN:
+			flag = luaW_toboolean(L, value.index)
+					? visibility::visible
+					: visibility::invisible;
+			break;
+		case LUA_TSTRING:
+			{
+				const std::string& str = lua_tostring(L, value.index);
+				if(str == "visible") {
+					flag = visibility::visible;
+				} else if(str == "hidden") {
+					flag = visibility::hidden;
+				} else if(str == "invisible") {
+					flag = visibility::invisible;
+				} else {
+					luaL_argerror(L, value.index, "string must be one of: visible, hidden, invisible");
+				}
+			}
+			break;
+		default:
+			luaW_type_error(L, value.index, "boolean or string");
+	}
+
+	w.set_visible(flag);
+
+	//if(flag == visibility::hidden) {
+	//	// HACK: this is needed to force the widget to be repainted immediately
```

Comment:
I'm assuming this is a hack that became unnecessary? In that case, why is it even still here?

Reply:
No i just commented it out to make the code easier for refactoring  i intended to check whether its still needed to bring it back in.  Also i'm not really a fan of unneeded invalidate_layout calls since they ,, well change the dialogs layout while its open which can be really confusing for the user.

I'd probably rather add a invalidate_layout function to the gui2 api. so that thare are less unconditional invaliadate_layout calls the the umc author cannot control.

Reply:
Adding a way to explicitly invalidate the layout in the GUI2 API seems alright to me. I'm not quite sure if I'd want to call it `invalidate_layout` but I can't think of anything better either, so I guess that'll do.

Reply:
> 'm not quite sure if I'd want to call it invalidate_layout but

Yes i agree but i cannot think of a better name